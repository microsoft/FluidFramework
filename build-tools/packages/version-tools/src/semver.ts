/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import * as semver from "semver";
import { VersionBumpTypeExtended } from "./bumpTypes";
import { bumpInternalVersion, getVersionRange } from "./internalVersionScheme";
import { adjustVersion, detectVersionScheme } from "./schemes";

/**
 * Return the version RANGE incremented by the bump type (major, minor, or patch).
 *
 * @remarks
 *
 * Only simple "^" and "~" ranges and Fluid internal version scheme ranges are supported.
 *
 * @param range - A dependency range string to increment.
 * @param bumpType - The type of bump.
 * @param prerelease - If true, will bump to a prerelease version.
 * @returns a bumped range string.
 */
export function incRange(
    range: string,
    bumpType: VersionBumpTypeExtended,
    prerelease = false,
): string {
    if (semver.validRange(range) === null) {
        throw new Error(`${range} is not a valid semver range.`);
    }

    const scheme = detectVersionScheme(range);
    switch (scheme) {
        case "virtualPatch":
        case "semver": {
            const operator = range.slice(0, 1);
            const isPreciseVersion = operator !== "^" && operator !== "~";
            const original = isPreciseVersion ? range : range.slice(1);
            const parsedVersion = semver.parse(original);
            const originalNoPrerelease = `${parsedVersion?.major}.${parsedVersion?.minor}.${parsedVersion?.patch}`;
            const newVersion =
                bumpType === "current"
                    ? originalNoPrerelease
                    : scheme === "virtualPatch"
                    ? adjustVersion(originalNoPrerelease, bumpType, "virtualPatch")
                    : semver.inc(originalNoPrerelease, bumpType);
            if (newVersion === null) {
                throw new Error(`Failed to increment ${original}.`);
            }
            return `${isPreciseVersion ? "" : operator}${newVersion}${prerelease ? "-0" : ""}`;
        }

        case "internal": {
            const constraintType = detectConstraintType(range);
            const original = semver.minVersion(range);
            if (original === null) {
                throw new Error(`Couldn't determine minVersion from ${range}.`);
            }
            const newVersion = bumpInternalVersion(original, bumpType);
            return getVersionRange(newVersion, constraintType);
        }

        default: {
            throw new Error(`${scheme} wasn't handled. Was a new version scheme added?`);
        }
    }
}

/**
 * Detects the type of upgrade constraint that a version range represents. Only works for Fluid internal version scheme
 * versions.
 *
 * @param range - The range to check.
 * @returns The constraint type.
 *
 * @remarks
 *
 * Throws an Error if the range is not valid.
 */
export function detectConstraintType(range: string): "minor" | "patch" {
    const minVer = semver.minVersion(range);
    if (minVer === null) {
        throw new Error(`Couldn't determine minVersion from ${range}.`);
    }

    const patch = bumpInternalVersion(minVer, "patch");
    const minor = bumpInternalVersion(minVer, "minor");

    const maxSatisfying = semver.maxSatisfying([patch, minor], range);
    return maxSatisfying === patch ? "patch" : "minor";
}
