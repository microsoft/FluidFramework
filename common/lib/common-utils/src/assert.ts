/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

/**
 * A browser friendly version of the node assert library. Use this instead of the 'assert' package, which has a big
 * impact on bundle sizes.
 * @param condition - The condition that should be true, if the condition is false an error will be thrown.
 * @param message - The message to include in the error when the condition does not hold.
 * A number should not be specificed manually: use a string.
 * Before a release, policy-check should be run, which will convert any asserts still using strings to
 * use numbered error codes instead.
 */
export function assert(condition: boolean, message: string | number): asserts condition {
    if (!condition) {
        throw new Error(
            typeof message === "number" ? `0x${message.toString(16).padStart(3, "0")}` : message,
        );
    }
}
