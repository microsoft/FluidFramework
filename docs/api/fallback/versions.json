{
	"currentVersion": "v2",
	"ltsVersion": "v1"
}
