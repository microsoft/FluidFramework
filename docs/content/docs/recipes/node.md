---
title: Using Fluid with NodeJS
menuPosition: 3
author: sdeshpande3
aliases:
  - "/start/nodejs-tutorial/"
---

In this tutorial, you'll learn about using the Fluid Framework by building a simple application in NodeJS that enables connected clients to generate random numbers and display the result of any changes to the shared state.  You'll also learn how to connect the Fluid data layer in [Node](https://nodejs.org/).

To jump ahead into the finished demo, check out the [Node demo in our FluidExamples repo](https://github.com/microsoft/FluidExamples/tree/main/examples/node-demo).

The following image shows the random number generated open in four terminals after every one second.

![Four terminals with the Random Number app open in them.](https://user-images.githubusercontent.com/46719950/140808244-abb4a046-6170-48e2-8766-c497e6ce977d.png)

{{< callout note >}}

This tutorial assumes that you are familiar with the [Fluid Framework Overview]({{< relref "/docs/_index.md" >}}) and that you have completed the [QuickStart]({{< relref "quick-start.md" >}}). You should also be familiar with the basics of [Node](https://nodejs.org/en/) and [creating Node projects](https://nodejs.org/en/about/).

{{< /callout >}}

## Create the project

1. Open a Command Prompt and navigate to the parent folder where you want to create the project; e.g., `c:\My Fluid Projects\fluid-node-tutorial`. The project is created in a subfolder named `fluid-node-tutorial`.

1. Run the following command at the prompt.

   ```dotnetcli
   npm init
   ```

1. Open the `package.json` file and replace the `main` and `scripts` sections with the following:

   ```json
   "main": "./src/index.js",
   "scripts": {
     "build": "npm build",
     "start:client": "node ./src/index.js",
     "start:server": "npx tinylicious@latest",
     "test": "echo \"Error: no test specified\" && exit 1",
     "test:report": "echo No test for this example"
   },
   ```

   And add the following line after the `dependencies` section, before the final closing bracket:

   ```json
   "type": "module"
   ```

1. The project uses two Fluid libraries:

    |Library |Description |
    |---|---|
    | `fluid-framework`    |Contains the SharedMap [distributed data structure]({{< relref "dds.md" >}}) that synchronizes data across clients. *This object will hold the most recent timestamp update made by any client.*|
    | `@fluidframework/tinylicious-client`   |Defines the connection to a Fluid service server and defines the starting schema for the [Fluid container][].|
    {.table}

    Run the following command to install the libraries.

    ```dotnetcli
    npm install @fluidframework/tinylicious-client fluid-framework readline-sync
    ```

{{< callout note >}}

You may need to install an additional dependency to make this demo compatible with Webpack 5. If you receive a compilation error related to a "buffer" package, please run `npm install -D buffer` and try again. This will be resolved in a future release of Fluid Framework.

{{< /callout >}}

## Code the project

1. Create `\src\index.js` file in your code editor and add the imports. The file should look like the following:

   ```js
   import { TinyliciousClient } from "@fluidframework/tinylicious-client";
   import { SharedMap } from "fluid-framework";
   import readlineSync from "readline-sync";
   ```

### Move Fluid data to the terminal

1. The Fluid runtime will bring changes made to the value from any client to the current client. Add the following code below the import statements.

    ```js

    // TODO 1: Configure the container

    function loadCli(map) {
       // TODO 2: Set the value that will appear on the terminal

       // TODO 3: Register handlers
   }

    async function createContainer() {
      // TODO 4: Create the container
    }

    async function loadContainer(id) {
      // TODO 5: Get the container from the Fluid service
    }

    async function start() {
        // TODO 6: Read container id from terminal and create/load container
    }

   start().catch(console.error());
    ```

1. Replace `TODO 1` with the following code. Note that there is only one object in the container: a SharedMap holding the random number. Note also that `sharedRandomNumber` is the ID of the `SharedMap` object and it must be unique within the container.

    ```js
      const client = new TinyliciousClient();
      const containerSchema = {
          initialObjects: { sharedRandomNumber: SharedMap }
      };
    ```

1. To ensure that both local and remote changes to the random number are reflected in the terminal, we will use the `newRandomNumber()` function to store the local value and `updateConsole()` function to ensure that the console is updated whenever any client changes the value. This helps in keeping the client terminals synchronized with the Fluid data.

   Replace `TODO 2` with the following code. It will set a timer to update the random number every one second.

   ```js
   const newRandomNumber = () => {
     map.set("random-Number-Key", Math.floor(Math.random() * 100) + 1);
   };
   setInterval(newRandomNumber, 1000);
   ```

  Replace `TODO 3` with the following code. It will listen for updates and print changes to the random number.

   ```js
   const updateConsole = () => {
     console.log("Value: ", map.get("random-Number-Key"));
   }
   updateConsole();
   map.on("valueChanged", updateConsole);
   ```

1. Replace `TODO 4` with the following code.

    ```js
    const { container } = await client.createContainer(containerSchema);
    container.initialObjects.sharedRandomNumber.set("random-Number-Key", 1);
    const id = await container.attach();
    console.log("Initializing Node Client----------", id);
    loadCli(container.initialObjects.sharedRandomNumber);
    return id;
    ```

1. Replace `TODO 5` with the following code.

   ```js
   const { container } = await client.getContainer(id, containerSchema);
   console.log("Loading Existing Node Client----------", id);
   loadCli(container.initialObjects.sharedRandomNumber);
   ```

2. Replace `TODO 6` with the following code. Note that, this code will first take the container id as the input. To create a new Fluid container, press Enter or type `undefined`. A new container will be initialized and the container id will be printed in the terminal. You can copy the container id, launch a new terminal window, and type/paste the initial container id to have multiple collaborative NodeJS clients.

   ```js
   const containerId = readlineSync.question("Type a Container ID or press Enter to continue: ");

   if(containerId.length === 0 || containerId === 'undefined' || containerId === 'null') {
     await createContainer();
   } else {
     await loadContainer(containerId);
   }
   ```

## Start the Fluid server and run the application

In the Command Prompt, run the following command to start the Fluid service. Note that `tinylicious` is the name of the Fluid service that runs on localhost.

   ```dotnetcli
   npx tinylicious@latest
   ```

Open a new Command Prompt and navigate to the root of the project; for example, `C:/My Fluid Projects/fluid-node-tutorial`. Start the application server with the following command. The application opens in your terminal.

   ```dotnetcli
   npm run start:client
   ```

To create a new Fluid container press Enter. The container id will be printed in the terminal. Copy the container id, launch a new terminal window, and type/paste the initial container id to have multiple collaborative NodeJS clients.

## Next steps

- You can find the completed code for this example in our Fluid Examples GitHub repository [here](https://github.com/microsoft/FluidExamples/tree/main/examples/node-demo).
- Try extending the demo with more key/value pairs and a more complex framework such as Express.
- Try changing the container schema to use a different shared data object type or specify multiple objects in `initialObjects`.

{{< callout tip >}}

When you make changes to the code the project will automatically rebuild and the application server will reload. However, if you make changes to the container schema, they will only take effect if you close and restart the application server. To do this, give focus to the Command Prompt and press Ctrl-C twice. Then run `npm run start:client` again.

{{< /callout >}}

<!-- AUTO-GENERATED-CONTENT:START (INCLUDE:path=docs/_includes/links.md) -->
<!-- Links -->

<!-- Concepts -->

[Fluid container]: {{< relref "containers.md" >}}

<!-- Distributed Data Structures -->

[SharedCounter]: {{< relref "/docs/data-structures/counter.md" >}}
[SharedMap]: {{< relref "/docs/data-structures/map.md" >}}
[SharedSequence]: {{< relref "/docs/data-structures/sequences.md" >}}
[SharedString]: {{< relref "/docs/data-structures/string.md" >}}

<!-- API links -->

[fluid-framework]: {{< relref "/docs/apis/fluid-framework.md" >}}
[@fluidframework/azure-client]: {{< relref "/docs/apis/azure-client.md" >}}
[@fluidframework/tinylicious-client]: {{< relref "/docs/apis/tinylicious-client.md" >}}

[AzureClient]: {{< relref "/docs/apis/azure-client/AzureClient-class.md" >}}
[TinyliciousClient]: {{< relref "/docs/apis/tinylicious-client/TinyliciousClient-class.md" >}}

[FluidContainer]: {{< relref "/docs/apis/fluid-static/fluidcontainer-class.md" >}}
[IFluidContainer]: {{< relref "/docs/apis/fluid-static/ifluidcontainer-interface.md" >}}

<!-- AUTO-GENERATED-CONTENT:END -->
