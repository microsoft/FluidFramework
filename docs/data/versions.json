{
	"params": {
		"currentVersion": "v2",
		"ltsVersion": "v1",
		"previousVersions": ["v1"],
		"downloadVersions": ["v1"]
	},
	"apiDocsDirectory": "/docs/api/",
	"apiVersionOptionName": {
		"Fluid Framework 1.0": "v1",
		"Fluid Framework 2.0": "v2"
	}
}
