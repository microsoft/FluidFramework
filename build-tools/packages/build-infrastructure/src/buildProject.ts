/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import path from "node:path";

import type { InterdependencyRange } from "@fluid-tools/version-tools";
import * as semver from "semver";
import { type SimpleGit, simpleGit } from "simple-git";

import { type BuildProjectConfig, getBuildProjectConfig } from "./config.js";
import { NotInGitRepository } from "./errors.js";
import { findGitRootSync } from "./git.js";
import {
	type IBuildProject,
	type IPackage,
	type IReleaseGroup,
	type IWorkspace,
	type PackageName,
	type ReleaseGroupName,
	type WorkspaceName,
} from "./types.js";
import { Workspace } from "./workspace.js";
import { loadWorkspacesFromLegacyConfig } from "./workspaceCompat.js";

/**
 * {@inheritDoc IBuildProject}
 */
export class BuildProject<P extends IPackage> implements IBuildProject<P> {
	/**
	 * The absolute path to the root of the build project. This is the path where the config file is located.
	 */
	public readonly root: string;

	/**
	 * {@inheritDoc IBuildProject.configuration}
	 */
	public readonly configuration: BuildProjectConfig;

	/**
	 * The absolute path to the config file.
	 */
	protected readonly configFilePath: string;

	/**
	 * @param searchPath - The path that should be searched for a BuildProject config file.
	 * @param gitRepository - A SimpleGit instance rooted in the root of the Git repository housing the BuildProject. This
	 * should be set to false if the BuildProject is not within a Git repository.
	 */
	public constructor(
		searchPath: string,

		/**
		 * {@inheritDoc IBuildProject.upstreamRemotePartialUrl}
		 */
		public readonly upstreamRemotePartialUrl?: string,
	) {
		const { config, configFilePath } = getBuildProjectConfig(searchPath);
		this.root = path.resolve(path.dirname(configFilePath));
		this.configuration = config;
		this.configFilePath = configFilePath;

		// Check for the buildProject config first
		if (config.buildProject === undefined) {
			// If there's no `buildProject` _and_ no `repoPackages`, then we need to error since there's no loadable config.
			if (config.repoPackages === undefined) {
				throw new Error(`Can't find configuration.`);
			} else {
				console.warn(
					`The repoPackages setting is deprecated and will no longer be read in a future version. Use buildProject instead.`,
				);
				this._workspaces = loadWorkspacesFromLegacyConfig(config.repoPackages, this);
			}
		} else {
			this._workspaces = new Map<WorkspaceName, IWorkspace>(
				Object.entries(config.buildProject.workspaces).map((entry) => {
					const name = entry[0] as WorkspaceName;
					const definition = entry[1];
					const ws = Workspace.load(name, definition, this.root, this);
					return [name, ws];
				}),
			);
		}

		const releaseGroups = new Map<ReleaseGroupName, IReleaseGroup>();
		for (const ws of this.workspaces.values()) {
			for (const [rgName, rg] of ws.releaseGroups) {
				if (releaseGroups.has(rgName)) {
					throw new Error(`Duplicate release group: ${rgName}`);
				}
				releaseGroups.set(rgName, rg);
			}
		}
		this._releaseGroups = releaseGroups;
	}

	private readonly _workspaces: Map<WorkspaceName, IWorkspace>;

	/**
	 * {@inheritDoc IBuildProject.workspaces}
	 */
	public get workspaces(): Map<WorkspaceName, IWorkspace> {
		return this._workspaces;
	}

	private readonly _releaseGroups: Map<ReleaseGroupName, IReleaseGroup>;

	/**
	 * {@inheritDoc IBuildProject.releaseGroups}
	 */
	public get releaseGroups(): Map<ReleaseGroupName, IReleaseGroup> {
		return this._releaseGroups;
	}

	/**
	 * {@inheritDoc IBuildProject.packages}
	 */
	public get packages(): Map<PackageName, P> {
		const pkgs: Map<PackageName, P> = new Map();
		for (const ws of this.workspaces.values()) {
			for (const pkg of ws.packages) {
				if (pkgs.has(pkg.name)) {
					throw new Error(`Duplicate package: ${pkg.name}`);
				}

				pkgs.set(pkg.name, pkg as P);
			}
		}

		return pkgs;
	}

	/**
	 * {@inheritDoc IBuildProject.relativeToRepo}
	 */
	public relativeToRepo(p: string): string {
		// Replace \ in result with / in case OS is Windows.
		return path.relative(this.root, p).replace(/\\/g, "/");
	}

	/**
	 * Reload the BuildProject by calling `reload` on each workspace in the repository.
	 */
	public reload(): void {
		for (const ws of this.workspaces.values()) {
			ws.reload();
		}
	}

	private gitRepository: SimpleGit | undefined;
	private _checkedForGitRepo = false;

	/**
	 * {@inheritDoc IBuildProject.getGitRepository}
	 */
	public async getGitRepository(): Promise<Readonly<SimpleGit>> {
		if (this.gitRepository !== undefined) {
			return this.gitRepository;
		}

		if (this._checkedForGitRepo === false) {
			this._checkedForGitRepo = true;
			// Check if the path is within a Git repo by trying to find the path to the Git repo root. If not within a git
			// repo, this call will throw a `NotInGitRepository` error.
			const gitRoot = findGitRootSync(this.root);
			this.gitRepository = simpleGit(gitRoot);
			return this.gitRepository;
		}

		throw new NotInGitRepository(this.root);
	}

	/**
	 * {@inheritDoc IBuildProject.getPackageReleaseGroup}
	 */
	public getPackageReleaseGroup(pkg: Readonly<P>): Readonly<IReleaseGroup> {
		const found = this.releaseGroups.get(pkg.releaseGroup);
		if (found === undefined) {
			throw new Error(`Cannot find release group for package: ${pkg}`);
		}

		return found;
	}
}

/**
 * Searches for a BuildProject config file and loads the project from the config if found.
 *
 * @typeParam P - The type to use for Packages.
 * @param searchPath - The path to start searching for a BuildProject config.
 * @param upstreamRemotePartialUrl - A partial URL to the upstream repo. This is used to find the local git remote that
 * corresponds to the upstream repo.
 * @returns The loaded BuildProject.
 */
export function loadBuildProject<P extends IPackage>(
	searchPath: string,
	upstreamRemotePartialUrl?: string,
): IBuildProject<P> {
	const repo = new BuildProject<P>(searchPath, upstreamRemotePartialUrl);
	return repo;
}

/**
 * Returns an object containing all the packages, release groups, and workspaces that a given set of packages depends
 * on. This function only considers packages in the BuildProject repo.
 */
export function getAllDependencies(
	repo: IBuildProject,
	packages: IPackage[],
): { packages: IPackage[]; releaseGroups: IReleaseGroup[]; workspaces: IWorkspace[] } {
	const dependencyPackages: Set<IPackage> = new Set();
	const releaseGroups: Set<IReleaseGroup> = new Set();
	const workspaces: Set<IWorkspace> = new Set();

	for (const pkg of packages) {
		for (const { name } of pkg.combinedDependencies) {
			const depPackage = repo.packages.get(name);
			if (depPackage === undefined) {
				continue;
			}

			if (pkg.releaseGroup !== depPackage.releaseGroup) {
				dependencyPackages.add(depPackage);
				releaseGroups.add(repo.getPackageReleaseGroup(depPackage));

				if (pkg.workspace !== depPackage.workspace) {
					workspaces.add(depPackage.workspace);
				}
			}
		}
	}

	return {
		packages: [...dependencyPackages],
		releaseGroups: [...releaseGroups],
		workspaces: [...workspaces],
	};
}

/**
 * Sets the dependency range for a group of packages given a group of dependencies to update.
 * The changes are written to package.json. After the update, the packages are
 * reloaded so the in-memory data reflects the version changes.
 *
 * @param packagesToUpdate - A list of objects whose version should be updated.
 * @param dependencies - A list of objects that the packagesToUpdate depend on that should have updated ranges.
 * @param dependencyRange - The new version range to set for the packageToUpdate dependencies.
 */
export async function setDependencyRange<P extends IPackage>(
	packagesToUpdate: Iterable<P>,
	dependencies: Iterable<P>,
	dependencyRange: InterdependencyRange,
): Promise<void> {
	const dependencySet = new Set(Array.from(dependencies, (d) => d.name));
	// collect the "save" promises to resolve in parallel
	const savePromises: Promise<void>[] = [];

	for (const pkg of packagesToUpdate) {
		for (const { name: depName, depKind } of pkg.combinedDependencies) {
			if (dependencySet.has(depName)) {
				const depRange =
					typeof dependencyRange === "string"
						? dependencyRange
						: dependencyRange instanceof semver.SemVer
							? dependencyRange.version
							: undefined;

				// Check if depRange is defined
				if (depRange === undefined) {
					throw new Error(`Invalid dependency range: ${dependencyRange}`);
				}

				// Update the version in packageJson
				if (depKind === "prod" && pkg.packageJson.dependencies !== undefined) {
					pkg.packageJson.dependencies[depName] = depRange;
				} else if (depKind === "dev" && pkg.packageJson.devDependencies !== undefined) {
					pkg.packageJson.devDependencies[depName] = depRange;
				} else if (depKind === "peer" && pkg.packageJson.peerDependencies !== undefined) {
					pkg.packageJson.peerDependencies[depName] = depRange;
				}
			}
		}
		savePromises.push(pkg.savePackageJson());
	}
	await Promise.all(savePromises);

	// Reload all packages to refresh the in-memory data
	for (const pkg of packagesToUpdate) {
		pkg.reload();
	}
}
