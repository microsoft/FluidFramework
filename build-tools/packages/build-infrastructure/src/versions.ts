/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import { SemVer } from "semver";

import { updatePackageJsonFileAsync } from "./packageJsonUtils.js";
import type { IPackage, PackageJson } from "./types.js";

/**
 * Sets the version of a group of packages, writing the new version in package.json. After the update, the packages are
 * reloaded so the in-memory data reflects the version changes.
 *
 * @param packages - An array of objects whose version should be updated.
 * @param version - The version to set.
 */
export async function setVersion<J extends PackageJson>(
	packages: IPackage<J>[],
	version: SemVer,
): Promise<void> {
	const translatedVersion = version;
	const setPackagePromises: Promise<void>[] = [];
	for (const pkg of packages) {
		setPackagePromises.push(
			updatePackageJsonFileAsync<J>(pkg.directory, async (json) => {
				json.version = translatedVersion.version;
			}),
		);
	}
	await Promise.all(setPackagePromises);

	// Reload all the packages to refresh the in-memory data
	for (const pkg of packages) {
		pkg.reload();
	}
}
