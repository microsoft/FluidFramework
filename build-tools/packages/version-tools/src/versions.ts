/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import child_process from "node:child_process";
import * as semver from "semver";

import { detectVersionScheme, getLatestReleaseFromList } from "./schemes";
import { changePreReleaseIdentifier, isInternalVersionScheme } from "./internalVersionScheme";

// TODO: Replace this with a shared release group type.
type TagPrefix = string | "client" | "server" | "azure" | "build-tools";

/**
 * Generates a simpler version scheme used for some packages and prereleases.
 */
export function getSimpleVersion(
	fileVersion: string,
	argBuildNum: string,
	argRelease: boolean,
	patch: boolean,
) {
	// Azure DevOp passes in the build number as $(buildNum).$(buildAttempt).
	// Get the Build number and ignore the attempt number.
	const buildId = patch ? parseInt(argBuildNum.split(".")[0], 10) : undefined;
	let version = fileVersion;
	if (isInternalVersionScheme(version, /* allowPrereleases */ true)) {
		if (patch) {
			throw new Error(
				`Cannot use simple patch versioning with Fluid internal versions. Version: ${fileVersion}`,
			);
		}

		if (!argRelease) {
			version = changePreReleaseIdentifier(version, "dev");
		}
	}

	const { releaseVersion, prereleaseVersion } = parseFileVersion(version, buildId);
	const build_suffix = buildId ? "" : argRelease ? "" : argBuildNum;
	const fullVersion = generateSimpleVersion(releaseVersion, prereleaseVersion, build_suffix);
	return fullVersion;
}

/**
 * @param prefix - The tag prefix to filter the tags by (client, server, etc.).
 * @param current_version - The version to test; that is, the version to check for being the latest build.
 * @param input_tags - If provided, only these tags will be considered. Mostly useful for testing.
 * @param includeInternalVersions - Whether to include Fluid internal builds, which are always
 * @param log - if true, logs to console.
 * @returns true if the current version is to be considered the latest (higher than the tagged releases _and NOT_ a
 * pre-release version).
 */
export function getIsLatest(
	prefix: TagPrefix,
	current_version: string,
	input_tags?: string[],
	includeInternalVersions = false,
	log = false,
): boolean {
	let latestTaggedRelease: string;

	if (input_tags?.length === 0) {
		latestTaggedRelease = "0.0.0";
	}

	let versions =
		input_tags === undefined ? getVersions(prefix) : getVersionsFromStrings(prefix, input_tags);
	versions = versions.filter((v) => {
		if (v === undefined) {
			return false;
		}

		if (includeInternalVersions) {
			return true;
		}

		return !isInternalVersionScheme(v);
	});

	latestTaggedRelease = getLatestReleaseFromList(versions);
	if (versions.length === 0 || latestTaggedRelease === undefined) {
		latestTaggedRelease = "0.0.0";
	}

	if (log) {
		console.log(`Latest tagged: ${latestTaggedRelease}, current: ${current_version}`);
	}
	const currentIsGreater = semver.gte(current_version, latestTaggedRelease);
	const currentIsPrerelease = includeInternalVersions
		? detectVersionScheme(current_version) === "internalPrerelease"
		: semver.prerelease(current_version) !== null;

	return currentIsGreater && !currentIsPrerelease;
}

/* A simpler CI version that append the build number at the end in the prerelease */
function generateSimpleVersion(
	release_version: string,
	prerelease_version: string,
	build_suffix: string,
) {
	// Generate the full version string
	if (prerelease_version) {
		if (build_suffix) {
			return `${release_version}-${prerelease_version}.${build_suffix}`;
		}
		return `${release_version}-${prerelease_version}`;
	}

	if (build_suffix) {
		return `${release_version}-${build_suffix}`;
	}

	return release_version;
}

function parseFileVersion(fileVersion: string, buildId?: number) {
	const split = fileVersion.split("-");
	let releaseVersion = split[0];
	split.shift();
	const prereleaseVersion = split.join("-");

	/**
	 * Use the build id for patch number if given
	 */
	if (buildId) {
		// split the prerelease out
		const r = releaseVersion.split(".");
		if (r.length !== 3) {
			console.error(`ERROR: Invalid format for release version ${releaseVersion}`);
			process.exit(9);
		}
		r[2] = (parseInt(r[2], 10) + buildId).toString();
		releaseVersion = r.join(".");
	}

	return { releaseVersion, prereleaseVersion };
}

/**
 * Extracts versions from the output of `git tag -l` in the working directory. The returned array will be sorted
 * ascending by semver version rules.
 *
 * @param prefix - The tag prefix to filter the tags by (client, server, etc.).
 * @returns An array of versions extracted from the output of `git tag -l`.
 */
function getVersions(prefix: TagPrefix) {
	const raw_tags = child_process.execSync(`git tag -l`, { encoding: "utf8" });
	const tags = raw_tags.split(/\s+/g).map((t) => t.trim());
	return getVersionsFromStrings(prefix, tags);
}

/**
 * Extracts versions from an array of strings, sorts them according to semver rules, and returns the sorted array.
 *
 * @param prefix - The tag prefix to filter the tags by (client, server, etc.).
 * @param tags - An array of tags as strings.
 * @returns An array of versions extracted from the provided tags.
 */
function getVersionsFromStrings(prefix: TagPrefix, tags: string[]) {
	const versions = tags
		.filter((v) => v.startsWith(`${prefix}_v`))
		.map((tag) => tag.substring(`${prefix}_v`.length));
	semver.sort(versions);
	return versions;
}
