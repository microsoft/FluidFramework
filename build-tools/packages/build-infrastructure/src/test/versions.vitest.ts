/*!
 * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
 * Licensed under the MIT License.
 */

import { strict as assert } from "node:assert";
import path from "node:path";

import * as semver from "semver";
import { simpleGit } from "simple-git";
import { afterEach, expect, describe, it } from "vitest";

import { loadBuildProject } from "../buildProject.js";
import type { ReleaseGroupName, WorkspaceName } from "../types.js";
import { setVersion } from "../versions.js";

import { pick, testDataPath, testRepoRoot } from "./init.js";

const repo = loadBuildProject(path.join(testDataPath, "./testRepo"));
const main = repo.releaseGroups.get("main" as ReleaseGroupName);
assert(main !== undefined);

const group2 = repo.releaseGroups.get("group2" as ReleaseGroupName);
assert(group2 !== undefined);

const group3 = repo.releaseGroups.get("group3" as ReleaseGroupName);
assert(group3 !== undefined);

const secondWorkspace = repo.workspaces.get("second" as WorkspaceName);
assert(secondWorkspace !== undefined);

/**
 * A git client rooted in the test repo. Used for resetting tests.
 */
const git = simpleGit(testRepoRoot);

describe("setVersion", () => {
	afterEach(async () => {
		await git.checkout(["HEAD", "--", testRepoRoot]);
		repo.reload();
	});

	it("release group", async () => {
		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		await setVersion(main.packages, semver.parse("1.2.1")!);

		const results = main.packages.map((pkg) => {
			return pick(pkg.packageJson, ["name", "version"]);
		});

		expect(results).toMatchSnapshot();
	});

	it("workspace", async () => {
		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		await setVersion(secondWorkspace.packages, semver.parse("2.2.1")!);

		const results = secondWorkspace.packages.map((pkg) => {
			return pick(pkg.packageJson, ["name", "version"]);
		});

		expect(results).toMatchSnapshot();
	});

	it("repo", async () => {
		const packages = [...repo.packages.values()];
		// eslint-disable-next-line @typescript-eslint/no-non-null-assertion
		await setVersion(packages, semver.parse("1.2.1")!);

		const results = packages.map((pkg) => {
			return pick(pkg.packageJson, ["name", "version"]);
		});

		expect(results).toMatchSnapshot();
	});
});
