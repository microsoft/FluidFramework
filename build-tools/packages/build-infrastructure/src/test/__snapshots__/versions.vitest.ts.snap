// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`setVersion > release group 1`] = `
[
  {
    "name": "main-release-group-root",
    "version": "1.2.1",
  },
  {
    "name": "pkg-a",
    "version": "1.2.1",
  },
  {
    "name": "pkg-b",
    "version": "1.2.1",
  },
  {
    "name": "@private/pkg-c",
    "version": "1.2.1",
  },
  {
    "name": "@shared/shared",
    "version": "1.2.1",
  },
]
`;

exports[`setVersion > repo 1`] = `
[
  {
    "name": "main-release-group-root",
    "version": "1.2.1",
  },
  {
    "name": "@group2/pkg-d",
    "version": "1.2.1",
  },
  {
    "name": "@group2/pkg-e",
    "version": "1.2.1",
  },
  {
    "name": "@group3/pkg-f",
    "version": "1.2.1",
  },
  {
    "name": "@group3/pkg-g",
    "version": "1.2.1",
  },
  {
    "name": "pkg-a",
    "version": "1.2.1",
  },
  {
    "name": "pkg-b",
    "version": "1.2.1",
  },
  {
    "name": "@private/pkg-c",
    "version": "1.2.1",
  },
  {
    "name": "@shared/shared",
    "version": "1.2.1",
  },
  {
    "name": "second-release-group-root",
    "version": "1.2.1",
  },
  {
    "name": "other-pkg-a",
    "version": "1.2.1",
  },
  {
    "name": "other-pkg-b",
    "version": "1.2.1",
  },
]
`;

exports[`setVersion > workspace 1`] = `
[
  {
    "name": "second-release-group-root",
    "version": "2.2.1",
  },
  {
    "name": "other-pkg-a",
    "version": "2.2.1",
  },
  {
    "name": "other-pkg-b",
    "version": "2.2.1",
  },
]
`;
